78/perform deposit withdraw

alfetopito opened the pull request:
Closes https://github.com/gnosis/dex-react/issues/78

- [x] Implemented `submitDeposit`
- [x] Implemented `submitWithdraw`

Not addressed:
- Confirmation modal when replacing already requested withdraw. To be done on https://github.com/gnosis/dex-react/issues/85
- Check for allowance on deposit. What should we do in this case? Submit new tx requesting user to increase allowance?

gnosis-info commented:
Travis automatic deployment:
 https://pr104--dexreact.review.gnosisdev.com

alfetopito commented on src/components/DepositWidget/Form.tsx:
```diff
@@ -147,7 +149,31 @@ export const Form: React.FC<FormProps> = (props: FormProps) => {
       // return !hasErrors
     }
-  }, [amountInput, validatorActive])
+  }, [amountInput, decimals, errors, totalAmount, validatorActive])
```
To be honest, all values were added by prettier plugin, using the react-hook rules.
Anything that's a state variable used inside the effect is put in the watchlist. One way to reduce this would be to move the function inside the effect.
Anyway, I remember Dmitry suggested to refactor this in a way that we don't need to apply the validation here, wasn't it?

As for the infinite loop, react-hooks are smart enough to not update the state if there are no changes https://reactjs.org/docs/hooks-faq.html#is-there-something-like-forceupdate

anxolin replied:
Yes, probably this is a smell of a not good practice. @DmitryBespalov would you like to make a proposal for this? If we don't address now this, we should review it together with #101 (kind of related)

> As for the infinite loop, react-hooks are smart enough to not update the state if there are no changes
Yep, but I guess if we change the errors (for other validation), then it triggers the effect, and then in case of an error, it changes errors again. It's true that in the second update it will stop :)

Velenir replied:
You're right. Infinite loop.
Also I'm concerned about using `totalAmount` and other BN values in any hook dependency as they are often recreated from string in the API, so new object each time, rerender each time.

But it can be addressed later when we have real contracts. Maybe caching BN creation or using `BN.toString()` in dependencies instead of raw BN.

anxolin replied:
But I think it's good that it re-evaluates the validation if `totalAmount` changes. For example, when pooling the new balances every 30s. I don't see a huge issue with that.

What I'm confused is why the prettier makes the effect depend on ALL the vars that it reads. 

Ahh, I just realised about the part that concerned me (the errors dependency in that hook). It will disapear if we don't use it inside, we pass a function to `setErrors` instead of using `errors` inside.

alfetopito commented on src/api/exchange/DepositApiMock.ts:
```diff
@@ -102,6 +102,10 @@ export class DepositApiMock implements DepositApi {
     pendingDeposits.batchId = currentBatchId
+
+    // mock transfer tokens from user's mock `wallet`
+    await this._erc20Api.transfer(tokenAddress, userAddress, this.getContractAddress(), amount)
```
Created an issue to address it https://github.com/gnosis/dex-react/issues/106

gnosis-info commented:
Travis automatic deployment:
 https://pr104--dexreact.review.gnosisdev.com

gnosis-info commented:
Travis automatic deployment:
 https://pr104--dexreact.review.gnosisdev.com

alfetopito commented:
> It has to do with the success of the two operations, I'm missing the part where:
> 
>     * It resets the form and close it
> 
>     * Notifies the row, so the row can highlight it for some seconds as in the other operations.


Addressed both in the latest commits.

Moved the `highlight` logic into a new hook.

gnosis-info commented:
Travis automatic deployment:
 https://pr104--dexreact.review.gnosisdev.com

gnosis-info commented:
Travis automatic deployment:
 https://pr104--dexreact.review.gnosisdev.com

Velenir commented on src/hooks/useHighlight.ts:
```diff
@@ -0,0 +1,30 @@
+  const triggerHighlight = (): void => {
+    if (mounted.current) {
+      setHighlight(true)
+      setInterval(() => {
```
I didn't get to check this PR in time, but...
Why `Interval`, why not `Timeout`
And if you really need Interval, need also to clean it up on unmount, as many times as you `triggerHighlight`, otherwise memory leak.

alfetopito replied:
My bad, I have been forgetting to wait for your input.

As for your suggestion, I'll create a new PR to address it.

anxolin replied:
Yep, let's try to get the 2 approvals, leaving some fair time for review.
We can always poke the person in the daily report or chat if we are waiting for a PR and we really want to merge

alfetopito replied:
https://github.com/gnosis/dex-react/pull/112
